packages/turbo/ios/RNVisitableView.swift

```diff
@@ -8,98 +8,86 @@
+    let configuration = WKWebViewConfiguration()
+    configuration.applicationNameForUserAgent = applicationNameForUserAgent as String?
+    return configuration
+  }()
+    
+  lazy var controller: RNVisitableViewController = {
+    let controller = RNVisitableViewController()
+    controller.delegate = self
+    return controller
+  }()
+    
+  override func didMoveToWindow() {
+    reactViewController().addChild(controller)
+    controller.view.frame = bounds // Fixes incorrect size of the webview
+    controller.didMove(toParent: reactViewController())
+    addSubview(controller.view)
   }
-
-  required init?(coder aDecoder: NSCoder) {
-    super.init(coder: aDecoder)
+    
+  func becameTopMostView() {
+    turboSession.delegate = self
   }
-  
-  public func getRNSesssion() -> RNSession? {
-    if (!Thread.isMainThread) {
-      print("getSession accessed from incorrect thread")
-      return nil
-    }
     
-    guard let sessionModule = self.bridge?.uiManager.moduleRegistry.module(forName: "RNVisitableViewModule") as? RNVisitableViewModule else {
-        print("Couldn't find session for sessionHandle:", sessionHandle ?? "default session")
-        return nil
+  public func handleMessage(message: WKScriptMessage){
+    if let messageBody = message.body as? [AnyHashable : Any] {
+      onMessage?(messageBody)
     }
-    return sessionModule.getSession(sessionHandle: sessionHandle)
   }
   
-  public func getSession() -> Session? {
-    return getRNSesssion()?.turboSession
+  public func injectJavaScript(code: NSString) -> Void {
+    webView.evaluateJavaScript(code as String)
   }
-  
-  override func didMoveToWindow() {
-
-    let url = URL(string: String(url))!
     
-    if (self.controller == nil) {
-      print("Open new VistableView with URL passed from JS: ", url)
-      setupViewController(url: url)
-      self.addSubview((controller?.view)!)
+  private func visit(){
+    if(controller.visitableURL?.absoluteString == url as String){
+      return
     }
+    performVisit()
   }
-  
-  func setupViewController(url: URL) {
-    self.controller = RNVisitableViewController(url: url)
-    self.controller?.delegate = self
-    self.reactViewController().addChild(self.controller!)
-    self.controller?.view.frame = bounds // Fixes incorrect size of the webview
-    self.controller?.didMove(toParent: self.reactViewController())
+    
+  public func performVisit(){
+    controller.visitableURL = URL(string: String(url))
+    turboSession.visit(controller)
   }
-  
-  func attachDelegateAndVisit(_ visitable: Visitable) {
-    let session = getSession()
-    session?.delegate = self
-    // Upon initial load, the session.visit function is called
-    // to set the currentVisit private variable.
-    if(session?.webView.url == nil){
-      session?.visit(visitable)
-      return
-    }
-    session?.visitableViewWillAppear(visitable)
+    
+  public func performRestorationVisit(){
+    turboSession.visit(controller, action: .restore)
```

Comment:
Have to call .restore, otherwise the previous screen scroll position resets to the top of the page.

Reply:
I'm wondering if this would be solved by calling `session.visitableViewWillAppear` for the new top-most view after dismissing a modal?

It looks like `Session` has the exact behavior:

https://github.com/hotwired/turbo-ios/blob/c9d060fcdec136c3808cd0764e956d375012c6fb/Source/Session/Session.swift#L225-L226

Probably requires that `topmostVisit` in `Session` would point to the modal's `Visit`, which I assume is what we'd want anyway.